inst/fit_model/fit_model.cpp

```diff
@@ -0,0 +1,223 @@
+    } else {
+      std::cout << "Failed to create output directory '" << output_abs << "'" << std::endl;
+    }
+  } else {
+    std::cout << "Writing to existing output directory " << output_abs << "'" << std::endl;
+  }
+
+  if (sim_years > 60) {
+    std::cout << "Running to max no of sim years: 60\n" << std::endl;
+    sim_years = 60;
+  }
+  if (hiv_steps_per_year > 10) {
+    std::cout << "Running max no of HIV steps per years: 10" << std::endl;
+    hiv_steps_per_year = 10;
+  }
+
+  const double dt = (1.0 / hiv_steps_per_year);
+  const int num_genders = 2;
+  const int age_groups_pop = 81;
+  const int fertility_first_age_group = 15;
+  const int age_groups_fert = 35;
+  const int disease_stages = 7;
+  const int treatment_stages = 3;
+  const int hiv_adult_first_age_group = 15;
+  const int adult_incidence_first_age_group = hiv_adult_first_age_group;
+  // Hardcoded 15-49 for now (35 groups within this band)
+  const int pAG_INCIDPOP = 35;
+  const int time_art_start = 30;
+
+  // Set working dir whilst reading files
+  std::filesystem::path old_dir = std::filesystem::current_path();
+  std::filesystem::current_path(input_abs);
+  auto exit_handler = [&old_dir]() {
+    std::filesystem::current_path(old_dir);
+  };
+//  std::atexit(exit_handler);
+
+  // Only fine-grained ages at first
+  leapfrog::Tensor1<int> age_groups_hiv_span_data = serialize::deserialize_tensor<int, 1>(
+      std::string("hAG_SPAN_full"));
+  const leapfrog::TensorMap1<int> age_groups_hiv_span = tensor_to_tensor_map<int, 1>(age_groups_hiv_span_data);
+  int age_groups_hiv = static_cast<int>(age_groups_hiv_span.size());
+  int age_groups_hiv_15plus = age_groups_hiv;
+  const int scale_cd4_mortality = 1;
+  int hIDX_15PLUS = 0;
+  const double art_alloc_mxweight = 0.2;
+
+  leapfrog::Tensor1<int> v = serialize::deserialize_tensor<int, 1>(std::string("artcd4elig_idx"));
+  for (int i = 0; i <= sim_years; ++i) {
+    // 0-based indexing in C++ vs 1-based indexing in R
+    v(i) = v[i] - 1;
+  }
+  const leapfrog::TensorMap1<int> artcd4elig_idx = tensor_to_tensor_map<int, 1>(v);
+
+  leapfrog::Tensor1<double> h(treatment_stages - 1);
+  for (int i = 0; i < treatment_stages - 1; ++i) {
+    h(i) = 0.5;
+  }
+  const leapfrog::TensorMap1<double> h_art_stage_dur = tensor_to_tensor_map<double, 1>(h);
+
+  leapfrog::Tensor2<double> base_pop_data = serialize::deserialize_tensor<double, 2>(
+      std::string("basepop"));
+  const leapfrog::TensorMap2<double> base_pop = tensor_to_tensor_map<double, 2>(base_pop_data);
+  leapfrog::Tensor3<double> survival_data = serialize::deserialize_tensor<double, 3>(
+      std::string("survival"));
+  const leapfrog::TensorMap3<double> survival = tensor_to_tensor_map<double, 3>(survival_data);
+  leapfrog::Tensor3<double> net_migration_data = serialize::deserialize_tensor<double, 3>(
+      std::string("net_migration"));
+  const leapfrog::TensorMap3<double> net_migration = tensor_to_tensor_map<double, 3>(net_migration_data);
```

Comment:
I don't think there is a way to do this in 1 is there? `params` object expects a TensorMap but I want this function to own the tensor data (as it has read it from disk). I tried putting this into a single helper which reads the data, creates the Tensor and returns a TensorMap over it but I think that won't work here as the Tensor data is then cleaned up when that helper function returns. So I think we have to do it in 2 bits like this?

I also tried putting all this into a helper function to then just return the `params` object but ran into the same problems with ownership there.